clusterloader/docs/vision.md

```diff
@@ -0,0 +1,318 @@
+of user-defined phases.
+The following subsections will describe this in a bit more details.
+
+
+### Tuning Set
+
+Since we would like to be fully load even very big clusters, we need to be able to
+create a number of “similar” objects. A “Tuning Set” concept will allow us to
+spread those operations across some time.
+We define Tuning Set as following:
+
+```
+struct TuningSet {
+	Name         string
+	InitialDelay time.Duration
+	// Exactly one of the following should be set.
+	QpsLoad        *QpsLoad
+	RandomizedLoad *RandomizedLoad
+	SteppedLoad    *SteppedLoad
+}
+
+// QpsLoad defines a uniform load with a given QPS.
+struct QpsLoad {
+	Qps float
+}
+
+// RandomizedLoad defines a load that is spread randomly
+// across a given total time.
+struct RandomizedLoad {
+	AverageQps float
+}
+
+// SteppedLoad defines a load that generates a burst of
+// a given size every X secods.
+struct SteppedLoad {
+	BurstSize int32
+	StepDelay time.Duration
+}
+```
+
+More policies can be introduced in the future.
+
+
+### Multiple copies of the same workload
+
+To fill-in large clusters, we would need spread objects across different namespaces.
+In many cases, it will be enough for users for many namespaces to contain the same
+objects (or to be more specific: objects created from the same templates). Obviously,
+we want the config for the test to be as small as possible.
+As a result, we will introduce the following rules:
+- In the top-level test definition, we will define the number of namespaces that will
+  be used by a test.
+- All the namespaces will have names of a form “namespace-<number>” for number
+  in range 1..N (where N is the number of namespaces in a test)
+
+To make it usable, we introduce the following type:
+
+```
+struct Range {
+	Min int32
+	Max int32
+}
+```
+
+In the context of namespaces, such object defines all namespaces "namespace-\<i\>" for
+`i` in the range [Min, Max].
+
+
+### Step
+
+Each step will consist of a number of create, update and delete operations (potentially
+on many different object types). At the high level, single step contains with a number
+of `Phase` definitions, which all are being run in parallel:
+
+```
+struct Step {
+	Phases []Phase
+}
+```
+
+#### Defining object type
+
+In order to update or delete an object, user need to be able to define type of
+object that this operation is about. Thus, we introduce  the following type for
+this purpose:
+
+```
+struct ObjectType {
+	APIGroup   string
+	APIVersion string
+	Kind       string
+}
+```
+
+Using this will allow us to easily use dynamic client in most of the places
+which may significantly simply Cluster Loader itself.
+
+
+#### Phase
+
+Phase declaratively defines state of objects we should reach in the underlying
+Kuberneces cluster. A single declaration may result in a number of create, update
+and delete operation depending on the current state of the cluster.
+
+We define the phase as following:
+
+```
+struct Phase {
```

Comment:
I can't think of a way in which we can use this way of defining Phase to for e.g doing sth like scale RCs from a given template to different values of `replicas` (like how we're doing in our load test).

So in the current form, IIUC this either means having 16k templates i.e one per RC (which I don't think we want to do) or scaling all RCs using a given template to the same no. of replicas (which can potentially change scenario of our load test).

Reply:
> which can potentially change scenario of our load test

Maybe that's fine, but I'm not sure.

Reply:
> I can't think of a way in which we can use this way of defining Phase to for e.g doing sth like scale RCs from a given template to different values of replicas (like how we're doing in our load test).

You can - the template will contain:
```
"numReplicas: "3+Rand()%5"
```

We don't need to change the scenario of our tests :)

Reply:
That's a cool idea! One thing to note however is this might mean evaluating the template 16k times (which is probably fine).

Reply:
Yes - that would require mutliple evaluations. I think it's fine, but it's a good point that it requires a note - adding.

Reply:
Added to caveats section.